December 2017

rezeau opened the pull request:
small additions january 2018

icc commented on js/dialogcards.js:
```diff
@@ -374,6 +565,13 @@ H5P.Dialogcards = (function ($, Audio, JoubelUI) {
+      'class': 'h5p-dialogcards-gotit truncated h5p-dialogcards-disabled',
+      'title': self.params.gotit
+    }).click(function () {
+          self.gotIt($(this).parents('.h5p-dialogcards-cardwrap'));
```
Too much indentation

icc commented on js/dialogcards.js:
```diff
@@ -489,13 +687,13 @@ H5P.Dialogcards = (function ($, Audio, JoubelUI) {
-      if (!$loadCard.length && self.$current.index() + 1 < self.params.dialogs.length) {
-        var $cardWrapper = self.createCard(self.params.dialogs[self.$current.index() + 1], self.$current.index() + 1)
+      if (!$loadCard.length && self.$current.index() + 1 < self.dialogs.length) {
+        var $cardWrapper = self.createCard(self.dialogs[self.$current.index() + 1])
```
The second parameter seems to be missing and it's required.

rezeau replied:
Nothing missing here. Please refer to lines 67-70 where self.params.dialogs is saved to self.dialogs for further use when save content state is used.

icc replied:
Hm, seems like `createCard()` takes three arguments, but only two are required:
```
C.prototype.createCard = function (card, cardNumber, setCardSizeCallback) {
```
This call only has one, and the call on _line 439_ appear to be incorrect as well.

rezeau replied:
You're right, I'll look into it.

rezeau replied:
OK, fixed. Because ALL cards are loaded at once, in function C.prototype.nextCard we can skip the Add next card routine.

icc commented on js/dialogcards.js:
```diff
@@ -489,13 +687,13 @@ H5P.Dialogcards = (function ($, Audio, JoubelUI) {
         self.resize();
       }
-
+      self.turnCardToFront ();
```
Whitespace between function name and parentheses

icc commented on js/dialogcards.js:
```diff
@@ -513,10 +711,43 @@ H5P.Dialogcards = (function ($, Audio, JoubelUI) {
       self.$current.removeClass('h5p-dialogcards-current');
       self.$current = $prev.addClass('h5p-dialogcards-current').removeClass('h5p-dialogcards-previous');
       self.setCardFocus(self.$current);
+      self.turnCardToFront ();
```
Whitespace between function name and parentheses

icc commented on js/dialogcards.js:
```diff
@@ -513,10 +711,43 @@ H5P.Dialogcards = (function ($, Audio, JoubelUI) {
+   * When navigating forward or backward, reset card to front view if has previously been turned
+   * so that user can see the Question side, not the Answer side of the card.
+   */
+
```
No empty lines between comment section and function declaration

icc commented on js/dialogcards.js:
```diff
@@ -909,6 +1145,150 @@ H5P.Dialogcards = (function ($, Audio, JoubelUI) {
+    /**
+   * Resets the task.
+   * Used in contracts.
+   * @public
```
`@public` isn't needed

icc commented on js/dialogcards.js:
```diff
@@ -513,10 +711,43 @@ H5P.Dialogcards = (function ($, Audio, JoubelUI) {
+            self.turnCard(self.$current);
+            var $cg = self.$current.find('.h5p-dialogcards-gotit');
+            $cg.addClass('h5p-dialogcards-disabled');
+        }
```
Too much indentation

icc commented on js/dialogcards.js:
```diff
@@ -909,6 +1145,150 @@ H5P.Dialogcards = (function ($, Audio, JoubelUI) {
+
+    var $feedback = $('<div>', {
+        'class': 'h5p-question-feedback-container'
+      }).appendTo(self.$inner);
```
Watch the indentation here

rezeau commented:
Fixed most of the requested (minor) changes requested. Remaining problems:
1.- _Should we use this.nbCards on line 483?_ I do not see this reference.
2.- _I believe completely removing the 'Load cards progressively' feature will cause trouble for some users as all the images and audios will be loaded before the content is even used, which can be a lot if you have ~50 cards._ I am well aware of this major change I introduced, but loading all cards rather than progressively was the only way I found for my "got it" feature to work. I will test this with a large amount of cards on my test site to find out what happens.
3.- _It looks like all cards are automatically turned back to the front now. Not sure if this will be OK with everyone._ Yes, I made this change on purpose because it looked more logical to me. Maybe we could ask users/testers what they think of the feature.
4.- The changing of the randomCards in semantics.json will require an upgrade script that maps the old true/false value to the new options. Yes, of course, but I do not understand well enough how the upgrade scripts work so I would leave it up to you once you agree to integrate my pull request in core.

**Other points to be dealt with:**
a. the changes in the _language files_. I do not know what the current process is which extracts (automatically or manually) the language strings from the semantics.json file into the .en.json file and the other lang files. Of course I can provide the French translations.

b. in the dialogcards.js file, // Set default behavior. some strings are outdated and do not seem to correspond to the reality of the dialogcards activity:
1. description: "Sit in pairs and make up sentences ...
2. dialogs: [{text: 'Horse',answer: 'Hest'},{text: 'Cow',answer: 'Ku'}]

c. The new features introduced in this version will have to pass the accessibility test.

icc commented:
1. Appears to have move to line 480, the code looks like this:
```
.replace('@total', (self.params.dialogs.length).toString());
```

rezeau commented on css/dialogcards.css:
```diff
@@ -318,6 +323,14 @@
+  -moz-box-shadow: inset 0 4px 0 #0e407a;
+  box-shadow: inset 0 4px 0 #0e407a;
+}
+
```
I thought I had added that prefix in my last commit. Will check.

rezeau commented:
1.- Should we use this.nbCards on line 483? I do not see this reference.
Appears to have move to line 480, the code looks like this:
.replace('@total', (self.params.dialogs.length).toString());
No, does not work.

icc commented:
Regarding my previous comment I've discussed them with **falcon-git** and concluded the following:

2. We cannot merge this in as long as it breaks the _Load cards progressively_ feature. Too many rely on it and will be displeased. There must be a way to not add all the cards all the time. Perhaps only up until the card you resume from?

3. This is UX and appears to be coupled to many of the tasks and texts referring to tasks. Adding this will break a lot of people's content. If this is a needed improvement I'm afraid it must be added as an option in the editor.

4. This is not too difficult you just need to add a function to the upgrades.js file for the next minor version, which in this case is 7, and we'll test it before releasing the next version:

```js
7: function (parameters, finished) {
  // Convert randomCards from boolean to string option
  if (parameters && parameters.behaviour && parameters.behaviour.randomCards !== undefined) {
    parameters.behaviour.randomCards = (parameters.behaviour.randomCards ? 'random' : 'normal')
  }

  // Complete
  finished(null, parameters);
}
```

Regarding your other comments:
a. We use the [H5P CLI](https://www.npmjs.com/package/h5p) tool to update and generate translation files. You can just leave it for now if you'd like and we'll make sure it's updated after merging.

b. Feel free to update these if you'd like. These are here to prevent the content type from crashing in case there's no input. (Like when inserting into Course Presentation)

c. I can test this when the rest is in place.

rezeau commented:
Thanks for your comments. The biggest (and I think only remaining) problem is n°2, the "Load cards progressively" feature.  I have created a DC activity with many cards & pics in it and it loads all cards without any delay, so I do not see this as a problem myself but I expect some DC users have created "unreasonably huge" activities with many (hundreds ?) of cards in it or they have a slow internet connexion so that they DO need this progressive load feature. I don't remember if my decision to remove this feature was prompted by problems with the "got it" feature or the "resume from saved state" features that I introduced. I will now take some time to look into the problem and hopefully come up with a solution.

I will first of all fix all the small remaining problems you mention.
I'll add a "turn cards to front when navigating" option to the editor parameters.
By the way, I did add the .h5p-dialogcards prefix to .h5p-joubelui-button:active in the CSS as you suggested and I don't know why it's not showing as _resolved_ above.

rezeau commented:
In my commit dated today 19 JAN 2018 I restored the "load cards progressively" feature, as requested by @icc . You wrote "_There must be a way to not add all the cards all the time. Perhaps only up until the card you resume from?_" Yes, that's what I'm doing. The good news is that upon resuming, any card which was previously OKed and then removed from the DOM is not loaded, so that only needed cards are loaded. I think that now my pull request is the most acceptable I can make it!

icc commented:
@rezeau The code looks a lot better now, and the upgrade.js code worked nicely. I've downloaded your branch and done some testing and unfortunately, I've run into some issues that we need to resolve before releasing your new additions: 

1. I get a JavaScript error when trying to use the Gotit button for the first time:
```
dialogcards.js:1228 Uncaught TypeError: Cannot read property 'splice' of undefined
```
It seems that `self.cardOrder` is only set when resuming previously used content, thus crashing for those who haven't tried the task before or for those who haven't enabled the resume feature. 
I tried adding an else and setting it to -1 but that didn't work. I'm not sure what the initial value should be in this case?

2. When testing on existing content the "gotit" feature is automatically enabled, it seems that everyone does not appreciate this since it changes the workflow of their existing tasks. I think the best way we can avoid this is by adding a checkbox for turning on the "gotit" feature.

3. For existing content the result page does not show the correct score, it only says '0/'.
It seems the `maxScore` is only available if I edit and save the content, or if I didn't change any of the behavior settings for the original content. I believe that by altering this line:
```js
self.params = $.extend({
```
into this:
```js
self.params = $.extend(true, {
```
the variables inside the `behaviour` object will always be copied if they don't exist. The original code just copies if the whole `behaviour` object doesn't exist. However, I suspect this can cause issues with the default dialogs:
```js
  dialogs: [
```
That remains to be seen but they might have to be moved into a separate `if` statement after the defaults are set.
Also, note that the scoring at the end must only be shown if the "gotit" feature is enabled.

5. The retry button at the end is always visible even if the `Enable "Retry" button` option isn't checked and that seems a bit weird.

rezeau commented:
@icc  thanks for your review
1. "_I get a JavaScript error when trying to use the Gotit button for the first time:_"
Easily fixed with this test: if (!$.isEmptyObject(this.cardOrder)) {self.cardOrder.splice(index, 1);}

2.- "_When testing on existing content the "gotit" feature is automatically enabled etc_"
I can easily add an "enable Got It" check box in semantics to turn that new feature ON/OFF. However, the score feature only makes sense (in my opinion) if the user reaches the end of the activity with all cards removed from the stack. In that case, how can we make the display of the score parameter dependent on the "got it" feature being enabled? Oh, I notice you mentioned this problem in your point n°3. I seem to remember that this question of making the display of one behaviour parameter dependent on the enabling of another parameter has been asked in the H5P forum but do not remember if a solution was provided. If not, then there should be a warning in the Score description "User will only get a score if the GotIt feature is enabled".
Also maybe a better name can be found for the "Got It" feature and should the "OK" button be re'named "Got It"?

3.- _For existing content the result page does not show the correct score, it only says '0/'.
It seems the maxScore is only available if I edit and save the content, or if I didn't change any of the behavior settings for the original content. I believe that by altering this line:_
Yes, I am aware of this problem but don't really know how to solve it. Can I leave it to you to try one of the solutions you suggest?

5.- _"The retry button at the end is always visible even if the Enable "Retry" button option isn't checked and that seems a bit weird."_
The problem here is that with the introduction of the "got It" feature there are now 2 "Retry" features. The original Retry button is displayed upon reaching the last card in the pack. It really means "re-start from the beginning". The Retry button I display after the user has reached the end of the pack after having removed ALL cards from the pack really means "reset" or start the game afresh. Should we call this "end-of-game" button **Reset** instead, similar to the Memory Game interface? In that case, should we add one more parameter as in MG "Add button for retrying when the game is over"?

icc commented:
2. Yes, all the new features with the exception of the resume capability must be optional, at least in the first version so we can see what people prefer before we change their defaults.
Adding a warning/description is fine for now, but it's possible to include the [H5P Editor Show When](https://github.com/h5p/h5p-editor-show-when) widget to hide an option.
I think "Got It" sounds fine for the button.

3. Perhaps it could be solved by doing a check to see if the maxScore is set and below 1. A score below 1 should probably be avoided since it can cause confusion with the result tracking software.

5. I don't think people will tell the difference between the buttons. It's probably safe to have the same option mean the retry button on the summary screen when the "Got It" feature is enabled and then mean the retry button at the last card if "Got It" is disabled.

rezeau commented:
@icc  thanks for your reply

2.- Agree with all you wrote. I will make the "got it" feature an option; add a warning to the Score option;  re-name the OK button "Got It".

3.- Well, _at the moment_, the Dialog Cards activity has 3 default "results" values:  Start time |  Points awarded | Max points awarded. It does not have a "Finish time" value because there is nothing to determine the end of the activity, of course. Points awarded and Max points awarded both have a value of 0, which is rather strange.
The Memory Game has a Start, Finish time, and a _default_ MaxScore and Score of 1 point.
I will try to do the masScore check that you suggest.

4. OK. I'll do that. Although---as a former teacher---I do not understand why a teacher would want to prevent a student to retry an activity in a learning context.

rezeau commented:
Hope my latest commit of today should meet all demands.

icc commented:
I think it works quite nicely now, and it's ready to be part of the next release!

rezeau commented:
Great! Thanks for your help and your patience with my amateurish coding.
